### PR TITLE
chore: add and use assertIsFoxEthStakingContractAddress type guard assertion and helper

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

- Use a type guard assertion to narrow to `FoxEthStakingContractAddress`
- Add a `getTypeGuardAssertion` convenience util to encourage more usage of this pattern moving forward

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A - codebase quality.

## Risk

Minimal - no run-time changes expected.

## Testing

Not applicable - no runtime changes expected.

### Engineering

☝️ 

### Operations

☝️ 

## Screenshots (if applicable)

N/A

### CODE REVIEW

File: src/lib/utils.ts [hunk 1]
```diff
@@ -105,3 +105,9 @@ export const deepUpsertArray = <T>(
   if (!level1) level1 = data[level1Key] = {}
   level1[level2Key] = union(level1[level2Key] ?? [], [value])
 }
+
```
Comment (0xApotheosis) [inline src/lib/utils.ts:108]: Handy little convenience function - just give it your type guard function and an assertion message and you get back a type guard assertion function.

File: src/lib/utils.ts [hunk 2]
```diff
@@ -105,3 +105,9 @@ export const deepUpsertArray = <T>(
...
+export const getTypeGuardAssertion =
+  <T, U>(typeGuard: (maybeT: T | U) => maybeT is T, message: string) =>
+  (value: T | U): asserts value is T => {
+    if (!typeGuard(value)) throw new Error(`${message}: ${value}`)
+  }
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/lib/utils.ts:113]: what is `U` used for here?

File: src/state/slices/opportunitiesSlice/constants.ts [hunk 3]
```diff
@@ -35,11 +36,18 @@ export const foxEthStakingContractAddresses = [
...
+const isFoxEthStakingContractAddress = (
+  address: FoxEthStakingContractAddress | string,
 ): address is FoxEthStakingContractAddress =>
   foxEthStakingContractAddresses.includes(address as FoxEthStakingContractAddress)
 
+export const assertIsFoxEthStakingContractAddress: (
+  value: FoxEthStakingContractAddress | string,
+) => asserts value is FoxEthStakingContractAddress = getTypeGuardAssertion(
+  isFoxEthStakingContractAddress,
+  "assertIsFoxEthStakingContractAddress: Contract address isn't a known ETH/FOX staking address",
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/constants.ts:48]: That should already be part of the call stack? If so, I'd remove it to keep the error simple but fine if we don't

File: src/lib/utils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/lib/utils.ts:113]: It's the type of `maybeT` if the type assertion fails.

E.g. in the below usage:

```ts
export const assertIsFoxEthStakingContractAddress: (
  value: FoxEthStakingContractAddress | string,
) => asserts value is FoxEthStakingContractAddress = getTypeGuardAssertion(
  isFoxEthStakingContractAddress,
  "assertIsFoxEthStakingContractAddress: Contract address isn't a known ETH/FOX staking address",
)
```

it enables the `value: FoxEthStakingContractAddress | string,` line to work: string becomes `U`.

**Review by gomesalexandre — DISMISSED** [review]
Discussed the `T | U` generic in-person, we will need it to assert this is the happy path. LGTM ser! 🍏 🧃

File: src/state/slices/opportunitiesSlice/constants.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/state/slices/opportunitiesSlice/constants.ts]: https://github.com/shapeshift/web/pull/3485/commits/a450f073ae08b363efca2aa6e4decac3c6558fa9

**Review by gomesalexandre — APPROVED** [review]

**Pull request merged**